internal/webhook/management_webhook.go

```diff
@@ -55,21 +58,109 @@ func (*ManagementValidator) ValidateCreate(_ context.Context, _ runtime.Object)
 }
 
 // ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
-func (v *ManagementValidator) ValidateUpdate(ctx context.Context, _, newObj runtime.Object) (admission.Warnings, error) {
+func (v *ManagementValidator) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) (admission.Warnings, error) {
 	const invalidMgmtMsg = "the Management is invalid"
 
-	mgmt, ok := newObj.(*hmcv1alpha1.Management)
+	newMgmt, ok := newObj.(*hmcv1alpha1.Management)
 	if !ok {
 		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected Management but got a %T", newObj))
 	}
 
+	oldMgmt, ok := oldObj.(*hmcv1alpha1.Management)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected Management but got a %T", oldObj))
+	}
+
+	if err := checkComponentsRemoval(ctx, v.Client, oldMgmt, newMgmt); err != nil {
+		return admission.Warnings{"Some of the providers cannot be removed"},
+			apierrors.NewInvalid(newMgmt.GroupVersionKind().GroupKind(), newMgmt.Name, field.ErrorList{
+				field.Forbidden(field.NewPath("spec", "providers"), err.Error()),
+			})
+	}
+
+	incompatibleContracts, err := getIncompatibleContracts(ctx, v, newMgmt)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %v", invalidMgmtMsg, err)
+	}
+
+	if incompatibleContracts != "" {
+		return admission.Warnings{"The Management object has incompatible CAPI contract versions in ProviderTemplates"}, fmt.Errorf("%s: %s", invalidMgmtMsg, incompatibleContracts)
+	}
+
+	return nil, nil
+}
+
+func checkComponentsRemoval(ctx context.Context, cl client.Client, oldMgmt, newMgmt *hmcv1alpha1.Management) error {
+	removedComponentNames := []string{}
+	for _, oldComp := range oldMgmt.Spec.Providers {
+		if !slices.ContainsFunc(newMgmt.Spec.Providers, func(newComp hmcv1alpha1.Provider) bool { return oldComp.Name == newComp.Name }) {
+			removedComponentNames = append(removedComponentNames, oldComp.Name)
+		}
+	}
+
+	if len(removedComponentNames) == 0 {
+		return nil
+	}
+
+	removedProvidersSet := make(map[string]struct{})
+	for _, m := range removedComponentNames {
```

Comment:
I don't think it makes much sense to iterate over the management object status. the diff from comparing two specs should be enough

Reply:
correct but it's unclear (to me) how is it possible to extract provider names having only component names (if templates are not presented, should we fetch a `releases` object then and try to find a template reference there?)

Reply:
yes, if not referenced directly, should be taken from the release. alternatively we'd need to ensure that the status represents the "old" version of the object by comparing generations (but in that case we'd have to fail validation if it doesn't)